File: backend/schemas/complaint.py
"""
schemas/complaint.py — HostelOps AI
======================================
Pydantic v2 schemas for the Complaint entity.
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator
from sqlalchemy.dialects.postgresql import UUID

from schemas.enums import (
    ClassifiedBy,
    ComplaintCategory,
    ComplaintSeverity,
    ComplaintStatus,
    OverrideReason,
)


class ComplaintCreate(BaseModel):
    text: str = Field(..., max_length=1000)
    is_anonymous: bool = False


class ComplaintRead(BaseModel):
    id: str
    student_id: str
    text: str
    is_anonymous: bool
    category: ComplaintCategory | None
    severity: ComplaintSeverity | None
    status: ComplaintStatus
    assigned_to: str | None
    confidence_score: float | None
    ai_suggested_category: ComplaintCategory | None
    ai_suggested_assignee: str | None
    requires_approval: bool
    classified_by: ClassifiedBy
    override_reason: OverrideReason | None
    flagged_input: str | None
    resolved_confirmed_at: datetime | None = None
    reopen_reason: str | None = None
    is_priority: bool = False
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

    @field_validator('id', 'student_id', 'assigned_to', 'ai_suggested_assignee', mode='before')
    @classmethod
    def uuid_to_str(cls, v):
        return str(v) if v is not None else None


class ComplaintReadAnonymous(BaseModel):
    """Safe schema for anonymous complaints — omits student_id."""
    id: str
    text: str
    is_anonymous: bool
    category: ComplaintCategory | None
    severity: ComplaintSeverity | None
    status: ComplaintStatus
    assigned_to: str | None
    confidence_score: float | None
    requires_approval: bool
    classified_by: ClassifiedBy
    resolved_confirmed_at: datetime | None = None
    reopen_reason: str | None = None
    is_priority: bool = False
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

    @field_validator('id', 'assigned_to', mode='before')
    @classmethod
    def uuid_to_str(cls, v):
        return str(v) if v is not None else None


class ComplaintUpdate(BaseModel):
    status: ComplaintStatus | None = None
    assigned_to: str | None = None
    override_reason: OverrideReason | None = None

    model_config = ConfigDict(from_attributes=True)


class ClassificationResult(BaseModel):
    """Returned by the LLM classification agent or fallback classifier."""
    category: ComplaintCategory
    severity: ComplaintSeverity
    classified_by: ClassifiedBy
    confidence_score: float = 0.0
    requires_approval: bool = False
    note: str | None = None
    suggested_assignee_role: str | None = None

File: frontend/src/types/complaint.ts
/**
 * types/complaint.ts — HostelOps AI
 * TypeScript types mirroring backend complaint schemas and enums.
 */

export type ComplaintCategory =
    | 'mess'
    | 'laundry'
    | 'maintenance'
    | 'interpersonal'
    | 'critical'
    | 'uncategorised';

export type ComplaintSeverity = 'low' | 'medium' | 'high';

export type ComplaintStatus =
    | 'INTAKE'
    | 'CLASSIFIED'
    | 'AWAITING_APPROVAL'
    | 'ASSIGNED'
    | 'IN_PROGRESS'
    | 'RESOLVED'
    | 'REOPENED'
    | 'ESCALATED';

export type ClassifiedBy = 'llm' | 'fallback' | 'manual' | 'warden_override';

export type OverrideReason =
    | 'wrong_category'
    | 'wrong_assignee'
    | 'wrong_severity'
    | 'other';

export type ApprovalStatus = 'pending' | 'approved' | 'corrected';

export interface ComplaintCreate {
    text: string;
    is_anonymous?: boolean;
}

export interface ComplaintRead {
    id: string;
    hostel_id: string;
    student_id: string;
    text: string;
    is_anonymous: boolean;
    category: ComplaintCategory | null;
    severity: ComplaintSeverity | null;
    status: ComplaintStatus;
    assigned_to: string | null;
    confidence_score: number | null;
    ai_suggested_category: ComplaintCategory | null;
    ai_suggested_severity: ComplaintSeverity | null;
    ai_suggested_assignee: string | null;
    requires_approval: boolean;
    classified_by: ClassifiedBy | null;
    classification_note: string | null;
    override_reason: OverrideReason | null;
    override_note: string | null;
    flagged_input: string | null;
    resolved_at: string | null;
    created_at: string;
    updated_at: string;
}

export interface ApprovalQueueItemRead {
    id: string;
    complaint_id: string;
    ai_suggested_category: ComplaintCategory;
    ai_suggested_severity: ComplaintSeverity;
    ai_suggested_assignee: string;
    confidence_score: number;
    status: ApprovalStatus;
    reviewed_by: string | null;
    override_reason: string | null;
    created_at: string;
    reviewed_at: string | null;
}
